Affix user dot on screen in user tracking mode

1ec5 opened the pull request:
Place the user dot at a fixed location on screen in user tracking mode, eliminating the jumping effect between user location updates. ~~In course tracking mode, weight the user puck down towards the bottom of the view, since it’s more important to see the road ahead than to see whoever’s tailgating you.~~ Added a new property to MGLMapView that controls the position of the user dot or user puck. By default, the user dot/puck is fixed in the center of the view, but it can be shifted to the top or bottom (e.g. for a GPS-style turn-by-turn view).

Conceptually, we now rely on entity interpolation to smooth out the large jumps between location updates. As in MapKit, the user dot may be up to one second behind where Core Location thinks the device is. When user tracking is initially enabled, MGLMapView “flies” to the first location reported by Core Location. This process may take longer than a second, but it gives the user more context as to their current location than the standard easing that MapKit performs.

* [x] Affix user dot
* [x] Fly to initial user location
* [x] Allow the user dot’s callout view to stay open (#2618, #3154)
* [x] Weight user puck
* [x] Make user dot location configurable

Depends on #3583. Fixes #1041 and fixes #2600. #1582 complements this PR by smoothing the user dot’s motion outside of user tracking mode, using view-based animation.

/cc @friedbunny @zugaldia

1ec5 commented on platform/ios/src/MGLMapView.mm:
```diff
@@ -3175,7 +3276,9 @@ - (void)unrotateIfNeededAnimated:(BOOL)animated
-    else if (self.direction < snapTolerance || self.direction > 360 - snapTolerance)
+    else if ((self.direction < snapTolerance || self.direction > 360 - snapTolerance)
+             && self.userTrackingMode != MGLUserTrackingModeFollowWithHeading
+             && self.userTrackingMode != MGLUserTrackingModeFollowWithCourse)
```
Note to self to remove this change once #3588 lands.

1ec5 commented:
Tentatively putting on 3.1.0, since it’s ready. Just needs lots of testing.

1ec5 commented on platform/ios/src/MGLMapView.mm:
```diff
@@ -3024,50 +3090,65 @@ - (void)locationManager:(__unused CLLocationManager *)manager didUpdateLocations
+                MGLMapCamera *camera = self.camera;
+                camera.centerCoordinate = self.userLocation.location.coordinate;
+                camera.heading = course;
+                camera.altitude = newLocation.horizontalAccuracy;
```
This approach seems to be more reliable, and it also works when the map is tilted, something unsupported by the old code, which was ported from the raster SDK, which didn’t support tilting.

friedbunny replied:
Happy that this is simplified! The camera movement is much slicker, too.

One continued quibble is that highly-accurate location fixes zoom in too far and context is lost. This is somewhat mitigated by the camera animation, but perhaps we could also adopt a minimum altitude?

We could also consider only zooming in if the map view's zoom level is relatively low; i.e., Apple Maps doesn't change the altitude/zoom unless the map is zoomed out to ~z8.

friedbunny replied:
I'm also seeing this zoom in beyond the accuracy ring, which is unintuitive.

1ec5 replied:
You're right, it does always seem to zoom into z18 for some reason. Perhaps it's because iOS-specific code assumes a field of view of 30 degrees, which is less than what some mbgl calculations assume.

1ec5 replied:
Fixed in dbfc363a7a58307ebcdf4696a9bca89d0401048d. I haven’t nailed down exactly what MapKit does, but it does seem to like z14 quite a lot.

friedbunny commented on platform/ios/src/MGLMapView.mm:
```diff
@@ -3003,10 +3069,10 @@ - (void)locationManager:(__unused CLLocationManager *)manager didUpdateLocations
-        if ([self.selectedAnnotation isEqual:self.userLocation]) [self deselectAnnotation:self.userLocation animated:NO];
+        if (self.userTrackingState != MGLUserTrackingStateBegan)
+        {
+            self.userLocation.location = newLocation;
```
This means we potentially drop a fresher location if the old transition is still in progress?

1ec5 replied:
Only for the transition to the initial location when user tracking is switched on. That transition may take longer than a second and shouldn't be interrupted.

friedbunny replied:
The first location is also likely to be relatively inaccurate and get preempted quickly by another update. Don't think this is a blocker here, just another reason to do #3610.

1ec5 replied:
MapKit does a simple, easeTo-style animation to the first location, which makes it easier to accommodate intervening updates. By contrast, this branch uses a Google-style flyTo, which would be easy to modify mid-flight even after implementing #3610.

In their own ways, both MapKit and Google Maps favor context and continuity over up-to-the-second precision. If a particular application needs to go immediately to the actual user location, it can pass NO into -setUserTrackingMode:animated:, which is being made public in this PR.

friedbunny commented:
Quick zoom doesn't appear to take into account the content insets, which means the user dot is moved during this gesture while user tracking is enabled... and then moved back to the proper placement afterwards.

![ezgif com-optimize](https://cloud.githubusercontent.com/assets/1198851/12418983/0158ee5e-be81-11e5-8446-ea5b967d6e8e.gif)

1ec5 commented:
Ah, good catch. All the gestures need to be updated to account for the content inset.

Additionally, heading changes, which come in much more frequently than location changes, tend to interrupt any running animation to the new location during heading tracking mode. The animation for heading changes needs to also pan to the current location, not just rotate.

1ec5 commented:
> Quick zoom and two-finger zoom do not appear to take into account the content insets

Fixed in d32e0e6ed7d1a27b31b374e34a6fac1ea48551ee. In that change, for MapKit parity, I also removed the logic that kicks you out of user tracking mode if your zoom gesture is centered somewhere other than the user dot.

1ec5 commented:
Failing builds are due to #1828.

1ec5 commented:
> Additionally, heading changes, which come in much more frequently than location changes, tend to interrupt any running animation to the new location during heading tracking mode.

This is also the current behavior. Fixing it will require #3625 and is out of scope for this PR.

friedbunny commented:
![YEAAAAAAAAAAH](https://cloud.githubusercontent.com/assets/1198851/12459035/a24d9c98-bf7a-11e5-876b-630346b5bd57.png)

1ec5 commented:
Everything passes; merging.
